### PR TITLE
[WIP] Add Distillation example for consistency models

### PR DESCRIPTION
**Pull request by ayushtues**
Adding an example to distill a diffusion model into a consistency model.

### CODE REVIEW

**Comment by williamberman** [comment]
re: https://github.com/huggingface/diffusers/pull/3492#issuecomment-1564175110

score estimate meaning the approximation of the score function i.e. the vector field pointing towards higher data density at a given noise level. This is the output of the network in the denoising score matching interpretation of diffusion networks. 

Intuitively, this is the incremental trajectory that the data follows as it is denoised, and the result of that incremental step is what is solved for. 

Formally, the training process takes a sample from the dataset, noises it, predicts the score, solves for the incrementally denoised image, and then trains the distilled model to predict the same initial image for both the noised and incrementally denoised image.

Yes, I think we could _try_ this in principle for any denoising process where we can get an X_0, X_t and X_{t-1}. I'd just be hesitant to stray too far from the original implementation as the argument for the loss function involves the error from the diffeq solver (admittedly the argument for the loss function is the part of the consistency model paper I have the least familiarity with)

**Comment by ayushtues** [comment]
Going on a family trip this week, will work on this from next week properly!

**Comment by ayushtues** [comment]
I think its best to work on this once the pipeline design and scheduler discussions are complete, and we know what goes where. Was trying to write a training script, but was unable to figure out what should go in the scheduler, what will stay in the pipeline etc.

Will help on reviewing and designing the pipeline for now.

**Comment by ayushtues** [comment]
@williamberman is it okay to keep a bunch of code for Heun Solver as standalone functions in the training script, which do not directly fit into existing schedulers, and are only useful for training consistency models? Like the Heun Solver they use, is somewhat different from the existing `HeunDiscreteScheduler`, and the important parts can be just abstracted out in a standalone function

**Comment by williamberman** [comment]
@ayushtues which parts, can you link me?

**Comment by ayushtues** [comment]
This is still a work in progress, but in the training example [here](https://github.com/dg845/diffusers/blob/8f858cb6d4d7f160e0b9378f7e331e2b689068d5/examples/consistency_models/train_consistency_distillation.py#LL595C1-L610C1), I am using Heun Solver to get a denoised image, which corresponds to the Heun Solver [here](https://github.com/openai/consistency_models/blob/6d26080c58244555c031dbc63080c0961af74200/cm/karras_diffusion.py#LL143C1-L160C27) in the original implementation, used [here](https://github.com/openai/consistency_models/blob/6d26080c58244555c031dbc63080c0961af74200/cm/karras_diffusion.py#L196) to get the denoised image

**Comment by williamberman** [comment]
ok will take a look later but in the meantime I'd just manually put what you need in the training script and then if we need to move it into the scheduler, we can do that after :)

**Comment by ayushtues** [comment]
Hi @williamberman @dg845 can you do a quick review of `train_consistency_distillation.py`?

File: examples/consistency_models/train_consistency_distillation.py [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+            for i in range(len(models)):
+                # pop models so that they are not loaded again
+                model = models.pop()
+
+                # load diffusers style into model
+                load_model = UNet2DModel.from_pretrained(input_dir, subfolder="unet")
+                model.register_to_config(**load_model.config)
+
+                model.load_state_dict(load_model.state_dict())
+                del load_model
```
Comment (williamberman) [inline examples/consistency_models/train_consistency_distillation.py:333]: You should be able to directly load weights into the model without loading twice

File: examples/consistency_models/train_consistency_distillation.py [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+        
+                )
+        target_model = UNet2DModel(
+                sample_size= args.resolution,
+                in_channels=3,
+                out_channels=3,
+                layers_per_block=2,
+                num_class_embeds=1000,
+                block_out_channels= [32, 64],
+                attention_head_dim=8,
```
Comment (williamberman) [inline examples/consistency_models/train_consistency_distillation.py]: I'm not sure what's going on here, why are we hardcoding model config?

File: examples/consistency_models/train_consistency_distillation.py [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+            import xformers
+
+            xformers_version = version.parse(xformers.__version__)
+            if xformers_version == version.parse("0.0.16"):
+                logger.warn(
+                    "xFormers 0.0.16 cannot be used for training in some GPUs. If you observe problems during training, please update xFormers to at least 0.0.17. See https://huggingface.co/docs/diffusers/main/en/optimization/xformers for more details."
+                )
+            model.enable_xformers_memory_efficient_attention()
+        else:
+            raise ValueError("xformers is not available. Make sure it is installed correctly")
```
Comment (williamberman) [inline examples/consistency_models/train_consistency_distillation.py]: I wouldn't include xformers in new training scripts, the default pytorch implementations of the kernels should be available

File: examples/consistency_models/train_consistency_distillation.py [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+    for epoch in range(first_epoch, args.num_epochs):
+        model.train()
+        progress_bar = tqdm(total=num_update_steps_per_epoch, disable=not accelerator.is_local_main_process)
+        progress_bar.set_description(f"Epoch {epoch}")
+        for step, batch in enumerate(train_dataloader):
+            # Skip steps until we reach the resumed step
+            if args.resume_from_checkpoint and epoch == first_epoch and step < resume_step:
+                if step % args.gradient_accumulation_steps == 0:
+                    progress_bar.update(1)
+                continue
```
Comment (williamberman) [inline examples/consistency_models/train_consistency_distillation.py]: This resuming to a point shouldn't be in the other training scripts as well sorry about that, if we need to start at a particular timestep, we can just start there. I believe controlnet does it correctly

File: examples/consistency_models/train_consistency_distillation.py (hunk 2 above)
Comment (ayushtues) [reply examples/consistency_models/train_consistency_distillation.py]: Following https://github.com/huggingface/diffusers/blob/f7cc9adc05c36985c8b3f3e95a54940ddac5b4ca/examples/unconditional_image_generation/train_unconditional.py#LL372C1-L398C48, we can either use a default configuration for our UNet, or as the user to give their own configuration for it

File: examples/consistency_models/train_consistency_distillation.py (hunk 1 above)
Comment (ayushtues) [reply examples/consistency_models/train_consistency_distillation.py:333]: I am not sure what you mean by loading twice?

File: examples/consistency_models/train_consistency_distillation.py (hunk 4 above)
Comment (ayushtues) [reply examples/consistency_models/train_consistency_distillation.py]: Changed to be similar to that of controlnet

File: examples/consistency_models/train_consistency_distillation.py (hunk 3 above)
Comment (ayushtues) [reply examples/consistency_models/train_consistency_distillation.py]: Removed!

File: examples/consistency_models/train_consistency_distillation.py [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+            # Sample noise that we'll add to the images
+            noise = torch.randn(clean_images.shape).to(clean_images.device)
+            # Sample a random timestep for each image, TODO - allow different timesteps in a batch
+            index = torch.randint(
+                0, noise_scheduler.config.num_train_timesteps-1,  (1,), device=clean_images.device
+            ).long()
+            timestep = timesteps[index]
+            timestep_prev = timestep + 1
+            # TODO, we should have an add noise in the scheduler maybe?
+            noised_image = clean_images + noise*append_dims(timestep, clean_images.ndim)
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: I will add a `add_noise` method in the scheduler that implements this step (e.g. sampling a noised version of the image $`x_{t_{n+ 1}}`$ from $`\mathcal{N}(x; t_{n + 1}^2\textbf{I})`$, following Line 4 of Algorithm 2).

File: examples/consistency_models/train_consistency_distillation.py [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+            scaled_timesteps = noise_scheduler.scale_timestep(timestep)
+            scaled_timesteps_prev = noise_scheduler.scale_timestep(timestep_prev) 
+            target_model_ema.copy_to(target_model.parameters())
+
+            with accelerator.accumulate(model):
+                # Predict the noise residual
+                model_output = model(noised_image, scaled_timesteps, class_labels=labels).sample
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: I think you need to call `noise_scheduler.scale_model_input(noised_image, timestep)` before giving the noisy image to the `model`. The `scale_model_input` method scales the model input by `c_in`, which corresponds to the `model_output = model(c_in * x_t, ...)` line in `denoise` in the original code (https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L347)

Note that `CMStochasticIterativeScheduler`'s `get_scalings_for_boundary_condition` method only calculates `(c_skip, c_out)` and not `(c_skip, c_out, c_in)` as in the [original code](https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L68-L78); `c_in` is calculated by `scale_model_input`, as described above.

File: examples/consistency_models/train_consistency_distillation.py [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+        for step, batch in enumerate(train_dataloader):
+            clean_images = batch["input"]
+            labels = batch["labels"]
+            # Sample noise that we'll add to the images
+            noise = torch.randn(clean_images.shape).to(clean_images.device)
+            # Sample a random timestep for each image, TODO - allow different timesteps in a batch
+            index = torch.randint(
+                0, noise_scheduler.config.num_train_timesteps-1,  (1,), device=clean_images.device
+            ).long()
+            timestep = timesteps[index]
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: I think `timestep` here should be `noise_scheduler.sigmas[index]` rather than `noise_scheduler.timesteps[index]`. My understanding is that whenever the paper talks about timesteps (e.g. $`\{\tau_n\}`$ in Algorithm 1 or $`\{t_n\}`$ in Algorithm 2) it's typically referring to timesteps already mapped to Karras sigmas (in decreasing and increasing order, respectively). It's confusing because the Karras sigma schedule can also be thought of as timesteps which break up the time interval $`[\epsilon, T] = [\sigma_{min}, \sigma_{max}]`$ into unequally sized subintervals that become smaller as we go backward in time (see Section 4 of the [paper](https://arxiv.org/pdf/2303.01469.pdf)).

File: examples/consistency_models/train_consistency_distillation.py [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+                model_output = model(noised_image, scaled_timesteps, class_labels=labels).sample
+                distiller = noise_scheduler.step(
+                    model_output, timestep, noised_image, use_noise=False
+                ).prev_sample
+
+                # Heun Solver to get previous timestep image using teacher model
+                # TODO - make this cleaner
+                samples = noised_image
+                x = samples
+                model_output = teacher_model(x, scaled_timesteps, class_labels=labels).sample
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: I think the teacher noised image `x` here should also be scaled appropriately with respect to how the teacher model (for example, `google/ddpm-cifar10-32`) is trained. When we get the teacher model

https://github.com/dg845/diffusers/blob/cd460ca2f5107ca6053e27c6e92f0aee2248627c/examples/consistency_models/train_consistency_distillation.py#L414

we should also be getting the teacher scheduler (e.g. `teacher_scheduler = DiffusionPipeline.from_pretrained(<teacher checkpoint>).scheduler`) and then calling its `scale_model_input` method, analogously to how the original implementation [gets the teacher diffusion with the teacher model](https://github.com/openai/consistency_models/blob/main/scripts/cm_train.py#L80-L82) and then [calls `teacher_diffusion`'s `denoise` method](https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L136-L140) to appropriately scale the teacher input.

Similarly, the teacher timestep should be scaled according to the teacher pipeline's scheduler, which is in general different from `scaled_timesteps`, which is scaled according to the student consistency model's scheduler.

File: examples/consistency_models/train_consistency_distillation.py [hunk 9]
```diff
@@ -0,0 +1,701 @@
...
+                teacher_denoiser = noise_scheduler.step(
+                    model_output, timestep, x, use_noise=False
+                ).prev_sample
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: Similarly to https://github.com/dg845/diffusers/pull/2/files#r1233164339, I believe this should use the teacher model's scheduler rather than `noise_scheduler`, which is the student consistency model's scheduler.

Alternatively, maybe we should not call `teacher_scheduler.step` and instead always take a Heun step to get the teacher `denoised_image` (as defined on Line 598 below). Not sure which one makes more sense; I believe that the EDM models used as the teacher models for the consistency distillation examples in the original code all happen to be trained with the Heun sampler as well. Section 4 in the [paper](https://arxiv.org/pdf/2303.01469.pdf) leads me to believe that any sampler can be used, not just the sampler the teacher model was trained on. @williamberman thoughts?

File: examples/consistency_models/train_consistency_distillation.py [hunk 10]
```diff
@@ -0,0 +1,701 @@
...
+                d = (x - teacher_denoiser) / append_dims(scaled_timesteps, x.ndim)
+                samples = x + d * append_dims(scaled_timesteps_prev - scaled_timesteps, x.ndim)
+                model_output = teacher_model(samples, scaled_timesteps_prev, class_labels=labels).sample
+                teacher_denoiser = noise_scheduler.step(
+                    model_output, timestep_prev, samples, use_noise=False
+                ).prev_sample
+
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: Same comment as https://github.com/dg845/diffusers/pull/2/files#r1233167689

File: examples/consistency_models/train_consistency_distillation.py [hunk 11]
```diff
@@ -0,0 +1,701 @@
...
+                # TODO - make this cleaner
+                samples = noised_image
+                x = samples
+                model_output = teacher_model(x, scaled_timesteps, class_labels=labels).sample
+                teacher_denoiser = noise_scheduler.step(
+                    model_output, timestep, x, use_noise=False
+                ).prev_sample
+                d = (x - teacher_denoiser) / append_dims(scaled_timesteps, x.ndim)
+                samples = x + d * append_dims(scaled_timesteps_prev - scaled_timesteps, x.ndim)
+                model_output = teacher_model(samples, scaled_timesteps_prev, class_labels=labels).sample
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: Same comment as https://github.com/dg845/diffusers/pull/2/files#r1233164339

File: examples/consistency_models/train_consistency_distillation.py [hunk 12]
```diff
@@ -0,0 +1,701 @@
...
+                teacher_denoiser = noise_scheduler.step(
+                    model_output, timestep_prev, samples, use_noise=False
+                ).prev_sample
+
+                next_d = (samples - teacher_denoiser) / append_dims(scaled_timesteps_prev, x.ndim)
+                denoised_image = x + (d + next_d) * append_dims((scaled_timesteps_prev - scaled_timesteps) /2, x.ndim)
+
+                # get output from target model
+                model_output = target_model(denoised_image, scaled_timesteps_prev, class_labels=labels).sample
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: Same comment as https://github.com/dg845/diffusers/pull/2/files#r1233160581

File: examples/consistency_models/train_consistency_distillation.py [hunk 13]
```diff
@@ -0,0 +1,701 @@
...
+                distiller_target = noise_scheduler.step(
+                    model_output, timestep_prev, denoised_image, use_noise=False
+                ).prev_sample
```
Comment (dg845) [inline examples/consistency_models/train_consistency_distillation.py]: I think the gradients should be turned off here, analogously to how the gradients are turned off in [`target_denoise_fn`](https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L127-L131). My understanding is that only the ("online") `model`''s parameters $`\theta`$ are updated using a backward pass, while the `target_model`'s parameters $`\theta^-`$ are only updated through the EMA update.

File: examples/consistency_models/train_consistency_distillation.py (hunk 5 above)
Comment (dg845) [reply examples/consistency_models/train_consistency_distillation.py]: Added.

File: examples/consistency_models/train_consistency_distillation.py (hunk 9 above)
Comment (ayushtues) [reply examples/consistency_models/train_consistency_distillation.py]: Hey @dg845 , I think the teacher model would _always_ use the Heun/Euler solver of the paper, since it is only used as a score function to provide the score `s_phi(x, t)`, and not as a model with its own independent sampler. My understanding is that it is similar to the classifiers used in classifier guidance, and we don't really care about how it was trained.

File: examples/consistency_models/train_consistency_distillation.py (hunk 8 above)
Comment (ayushtues) [reply examples/consistency_models/train_consistency_distillation.py]: For scaling, I think yeah it might make sense to use the scheduler of the teacher, need to investigate this more

**Comment by dg845** [comment]
@ayushtues since the `consistency-model-pipeline` branch has been merged, it might make sense to open this as a PR directly in the `huggingface/diffusers` repo.

**Comment by ayushtues** [comment]
Yes makes sense, will open a new PR to Diffusers directly.

**Comment by ayushtues** [comment]
Started a new PR for this https://github.com/huggingface/diffusers/pull/3992
